src/layer_builder.rs

```diff
@@ -0,0 +1,278 @@
+use bevy::{
+    prelude::*,
+    render::{
+        mesh::{Indices, VertexAttributeValues},
+        pipeline::PrimitiveTopology,
+    },
+};
+
+/// Useful for creating and modifying a layer in the same system.
+pub struct LayerBuilder<T> {
+    pub settings: LayerSettings,
+    pub(crate) tiles: Vec<(Entity, Option<(T, bool)>)>,
+    pub(crate) layer_entity: Entity,
+}
+
+impl<T> LayerBuilder<T>
+where
+    T: TileBundleTrait,
+{
+    /// Creates the layer builder using the layer settings.
+    pub fn new(commands: &mut Commands, layer_entity: Entity, settings: LayerSettings) -> Self {
+        let tile_size_x = round_to_power_of_two((settings.map_size.x * settings.chunk_size.x) as f32);
+        let tile_size_y = round_to_power_of_two((settings.map_size.y * settings.chunk_size.y) as f32);
+        let tile_count = tile_size_x * tile_size_y;
+        Self {
+            settings,
+            tiles: (0..tile_count)
+                .map(|_| {
+                    //let mut tile_entity = None;
+                    // commands.entity(layer_entity).with_children(|child_builder| {
+                    // Commented out because child tiles cut performance in half.
+                    let tile_entity = Some(commands.spawn().id());
+                    // });
+                    (tile_entity.unwrap(), None)
+                })
+                .collect(),
+            layer_entity,
+        }
+    }
+
+    /// Sets a tile's data at the given position.
+    pub fn set_tile(
+        &mut self,
+        tile_pos: UVec2,
+        tile: T,
+        visible: bool,
+    ) -> Result<(), MapTileError> {
+        let morton_tile_index = morton_index(tile_pos);
+        if morton_tile_index < self.tiles.capacity() {
+            self.tiles[morton_tile_index].1 = Some((tile, visible));
+            return Ok(());
+        }
+        Err(MapTileError::OutOfBounds)
+    }
+
+    /// Returns a tile entity.
+    pub fn get_tile_entity(&mut self, tile_pos: UVec2) -> Result<Entity, MapTileError> {
+        let morton_tile_index = morton_index(tile_pos);
+        if morton_tile_index < self.tiles.capacity() {
+            return Ok(self.tiles[morton_tile_index].0);
+        }
+
+        Err(MapTileError::OutOfBounds)
+    }
+
+    /// Gets a reference to the tile data using the a tile position.
+    pub fn get_tile(&self, tile_pos: UVec2) -> Result<&T, MapTileError> {
+        let morton_tile_index = morton_index(tile_pos);
+        if morton_tile_index < self.tiles.capacity() {
+            if let Some(tile) = &self.tiles[morton_tile_index].1 {
+                return Ok(&tile.0);
+            } else {
+                return Err(MapTileError::NonExistent);
+            }
+        }
+        Err(MapTileError::OutOfBounds)
+    }
+
+    fn get_tile_i(&self, tile_pos: IVec2) -> Option<(bevy::prelude::Entity, &T)> {
```

Comment:
I think it would be nicer to name this `get_tile_i32()` (if we want to stick with glam's naming style, like `as_i32()` for example) or get_tile_ivec(), `i` is always used to name indexes so the function name is a bit confusing..